File: 0138-copy-list-with-random-pointer/0138-copy-list-with-random-pointer.js
/**
 * // Definition for a _Node.
 * function _Node(val, next, random) {
 *    this.val = val;
 *    this.next = next;
 *    this.random = random;
 * };
 */

/**
 * @param {_Node} head
 * @return {_Node}
 */

 /*
1. use 2 passes
in 1st pass, construct hashMap of all the copies
in 2nd pass assign pointers
2. hashMap = {null: null} // to handle edge case
 */
var copyRandomList = function(head) {
    var hashMap = new Map()
    var curr = head

    hashMap.set(null, null)

    while(curr) {
        var newNode = new _Node(curr.val, curr.next, curr.random)
        hashMap.set(curr, newNode)
        curr = curr.next
    }

    curr = head
    while(curr) {
        var newNode = hashMap.get(curr)
        newNode.next = hashMap.get(curr.next)
        newNode.random = hashMap.get(curr.random)
        curr = curr.next
    }
    return hashMap.get(head)
};

File: 0138-copy-list-with-random-pointer/0138-copy-list-with-random-pointer.py
"""
# Definition for a Node.
class Node:
    def __init__(self, x: int, next: 'Node' = None, random: 'Node' = None):
        self.val = int(x)
        self.next = next
        self.random = random
"""

class Solution:
    def copyRandomList(self, head: 'Optional[Node]') -> 'Optional[Node]':
        nodeMap = {None: None}

        curr = head
        while curr:
            nodeMap[curr] = Node(curr.val, curr.next, curr.random)
            curr = curr.next
        
        curr = head
        while curr:
            newNode = nodeMap.get(curr)
            newNode.next = nodeMap.get(curr.next)
            newNode.random = nodeMap.get(curr.random)
            curr = curr.next
        
        return nodeMap[head]